### PR TITLE
Implement CoNECo dataset

### PR DESCRIPTION
**Pull request by oguzserbetci**
- **Name:** CoNECo
- **Description:** Complex Named Entity Corpus (CoNECo) is an annotated corpus for NER and NEN of protein-containing complexes. CoNECo comprises 1,621 documents divided into train, val and test splits with 2,052 entities, 1,976 of which are normalized to Gene Ontology.
- **Paper:** https://www.biorxiv.org/content/early/2024/05/29/2024.05.18.594800
- **Data:** https://zenodo.org/records/11263147

### Checkbox

- [ ] ~~Confirm that this PR is linked to the dataset issue.~~
- [x] Create the dataloader script `hub/hub_repos/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [ ] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio_hub <dataset_name> [--data_dir /path/to/local/data] --test_local`.
- [ ] ~~If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.~~

### CODE REVIEW

**Review by phlobo — CHANGES_REQUESTED** [review]
Hi @oguzserbetci, thank you for the PR! 

Apart from the comments, I wondered if you have any idea why the entity counts are so much different from the paper - for instance, the paper reports 1360 mentions in the training set, but here we have 1589 in the training split.

Are there any annotations (e.g., duplicates), which are removed in the paper?

```
train
==========
id: 983
document_id: 983
passages: 983
entities: 1589
events: 0
coreferences: 0
relations: 0
normalized: 1310
```

File: bigbio/hub/hub_repos/coneco/coneco.py
[the remark this answers is not in the document]
Comment (phlobo) [reply bigbio/hub/hub_repos/coneco/coneco.py:1]: (also, _PUBMED)

File: bigbio/hub/hub_repos/coneco/coneco.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                    "filepath": data_dir / "dev",
+                    "split": "dev",
+                },
+            ),
+        ]
+
+    def _generate_examples(self, filepath, split: str) -> Tuple[int, Dict]:
+        """Yields examples as (key, example) tuples."""
+        if self.config.schema == "source":
+            for file in filepath.iterdir():
```
[the remark this answers is not in the document]
Comment (oguzserbetci) [reply bigbio/hub/hub_repos/coneco/coneco.py:173]: Great tip thanks!

**Comment by oguzserbetci** [comment]
> Hi @oguzserbetci, thank you for the PR!
> 
> Apart from the comments, I wondered if you have any idea why the entity counts are so much different from the paper - for instance, the paper reports 1360 mentions in the training set, but here we have 1589 in the training split.
> 
> Are there any annotations (e.g., duplicates), which are removed in the paper?
> 
> ```
> train
> ==========
> id: 983
> document_id: 983
> passages: 983
> entities: 1589
> events: 0
> coreferences: 0
> relations: 0
> normalized: 1310
> ```

So I have been looking into this but without any solution. I counted entities in the brat files and also the conll files, and they are close to 1589 albeit not consistent... I couldn't read anything from the paper. Initial version of the dataset on Zededo has the same number of entities.

File: bigbio/hub/hub_repos/coneco/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+- **Pubmed:** False
+- **Public:** True
+- **Tasks:** NER, NEN
+
+Complex Named Entity Corpus (CoNECo) is an annotated corpus for NER and NEN of protein-containing complexes. CoNECo comprises 1,621 documents with 2,052 entities, 1,976 of which are normalized to Gene Ontology. We divided the corpus into training, development, and test sets.
+
+## Citation Information
+
+```
+@article {Nastou2024.05.18.594800,
```
[the remark this answers is not in the document]
Comment (phlobo) [reply bigbio/hub/hub_repos/coneco/README.md:34]: @oguzserbetci could you please include the same citation in Readme and code?

File: bigbio/hub/hub_repos/coneco/coneco.py
[the remark this answers is not in the document]
Comment (phlobo) [reply bigbio/hub/hub_repos/coneco/coneco.py:1]: Unfortunately, tests are still failing:

```
raise AssertionError(f"Dataloader attribute '{metadata_attr}' not valid for {metadata_name} must be one of {license_keys}")
AssertionError: Dataloader attribute 'Creative Commons Attribution 4.0 International' not valid for _LICENSE must be one of {
```

File: bigbio/hub/hub_repos/coneco/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (oguzserbetci) [reply bigbio/hub/hub_repos/coneco/README.md]: opps, done :)

File: bigbio/hub/hub_repos/coneco/coneco.py
[the remark this answers is not in the document]
Comment (oguzserbetci) [reply bigbio/hub/hub_repos/coneco/coneco.py:1]: Oh ok, got this now too.

**Comment by oguzserbetci** [comment]
I checked again, there are no overflowing or duplicate offset entities. Also I was able to locate entity mentions in the text. Here is how I tested:
```python
>>> all([np.max([np.max(ent['offsets']) for ent in sample['entities']]) <= np.max([np.max(p['offsets']) for p in sample['passages']]) for sample in data['train'] if sample['entities']])
True
>>> all([sample['passages'][0]['text'][0][ent_begin:ent_end] == ent_text for sample in data['train'] for ent in sample['entities'] for (ent_begin, ent_end), ent_text in zip(ent['offsets'], ent['text'])])
True
```

File: bigbio/hub/hub_repos/coneco/coneco.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+CoNECo is an annotated corpus for NER and NEN of protein-containing complexes. CoNECo comprises 1,621 documents with 2,052 entities, 1,976 of which are normalized to Gene Ontology. We divided the corpus into training, development, and test sets.
+"""
+
+from pathlib import Path
+from typing import Dict, List, Tuple
+
+import datasets
+
+from bigbio.utils.license import Licenses
```
[the remark this answers is not in the document]
Comment (phlobo) [reply bigbio/hub/hub_repos/coneco/coneco.py:29]: @oguzserbetci please also remove the unused import.

Flake output:
```
bigbio/hub/hub_repos/coneco/coneco.py:29:1: F401 'bigbio.utils.license.Licenses' imported but unused
bigbio/hub/hub_repos/coneco/coneco.py:59:120: E501 line too long (131 > 119 characters)
```

**Comment by phlobo** [comment]
> The number of normalized entity mentions do equate to the paper. It is curious indeed. Should I contact the paper authors?

@oguzserbetci: that would be amazing! :)

**Comment by oguzserbetci** [comment]
@phlobo I contacted an author and got the issue resolved. There were entities labeled as out-of-scope, which I now filter out and resulting entity numbers match the paper.

```
train
==========
id: 983
document_id: 983
passages: 983
entities: 1360
normalized: 1310
events: 0
coreferences: 0
relations: 0

test
==========
id: 318
document_id: 318
passages: 318
entities: 283
events: 0
coreferences: 0
relations: 0
normalized: 279

validation
==========
id: 320
document_id: 320
passages: 320
entities: 409
normalized: 387
events: 0
coreferences: 0
relations: 0
```

image001

**Review by phlobo — APPROVED** [review]
@oguzserbetci Amazing, thanks for resolving all issues and doing the extra research about OOS entities!

**Pull request merged**